### ISSUE TITLE
[hip-runtime-amd/rocm-llvm] Conflicting definition of `__noinline__` with GCC 12.1

### ISSUE THREAD

**Issue by tpkessler** [rocm-arch/rocm-arch#782]
Fixed with https://github.com/rocm-arch/rocm-arch/commit/bffaf315960589541ca093671fa389ce9c9d365b, relevant upstream issue https://github.com/ROCm-Developer-Tools/HIP/issues/2678

**Comment by PlasmaPower** [comment]
FYI this seems to break `ptorch` rocm builds which now error complaining that `__noinline__` is an unknown type. Adding:
```
#if defined(__GNUC__)
#define __noinline__
#else
#define __noinline__ __attribute__((noinline))
#endif // !__GNUC__
#define __forceinline__ inline __attribute__((always_inline))
```
to `/opt/rocm/hip/include/hip/amd_detail/host_defines.h` just after the `#endif // !__CLANG_HIP_RUNTIME_WRAPPER_INCLUDED__` fixes that issue locally, which I took from https://github.com/ROCm-Developer-Tools/HIP/issues/2677#issuecomment-1126775021

**Comment by tpkessler** [comment]
Hey @PlasmaPower thanks for bringing up this problem! I've decided to fix this issue by patching `rocm-llvm` from the corresponding upstream issue, https://reviews.llvm.org/D124866
pytorch should work after you have built the new `rocm-llvm`.

**Comment by tpkessler** [comment]
In summary:
Fixed by modifying
`rocm-llvm` in 5ebd28cdd7b79c043767e39d6496b54e24bca543
`hip-rocm-runtime` in bffaf315960589541ca093671fa389ce9c9d365b

**Comment by tpkessler** [comment]
Patched upstream with 5.3.0.
